File: Parallel-Csharp/src/Parallel/Containers/VectorParallel.cs
namespace Parallel_Csharp.Parallel.Containers;

public class VectorParallel {
    // T1
    public List<char> GetListChar(){
        List<char> vec1 = new List<char> { 'a','b','c'};
        vec1.Add('d');
        return vec1;
    }

    public List<char> GetNCopyChar(char ele, int size){
        List<char> vec = new List<char>(Enumerable.Repeat(ele, size));
        return vec;
    }

    // T2
    public void ReSizeListChar(List<char> vec, int size, char ele){
        if (vec.Count >= size) {
            vec.RemoveRange(size, vec.Count);
        } else {
            vec.AddRange(Enumerable.Repeat(ele, size - vec.Count));
        }
    }

    // T3
    public void ClearSizeListChar(List<char> vec, int size){
        if(vec.Count == 0) return;
        if(vec.Count <= size) {
            vec.Clear();
        }
    }

    // T4
    public int InsertListChar(List<char> vec, int index, char ele){
        if(index > vec.Count || index < 0) return -1;
        vec.Insert(index, ele);
        return vec.Count;
    }

    // T5
    public int InsertListCharWithList(List<char> vec, int index, List<char> eles){
        if(index > vec.Count || index < 0) return -1;
        vec.InsertRange(index, eles);
        return vec.Count;
    }

    // T6
    public bool ContainsAnyChar(List<char> vec, List<char> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(char ele in eles){
            if(vec.Contains(ele))
                return true;
        }
        return false;
    }

    public bool ContainsAllChar(List<char> vec, List<char> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(char ele in eles){
            if(!vec.Contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void SetListChar(List<char> vec, int position , char v){
        if(position < 0 || position >= vec.Count) return;
        else{
            vec[position] = v;
        }
    }

    public void SetListCharWithList(List<char> vec, List<int> position, List<char> v){
        if(vec.Count < position.Count) return;
        else{
            for(int i=0; i<position.Count; ++i){
                vec[position[i]] = v[i];
            }
        }
    }

    // T8
    public void RemoveListChar(List<char> vec, int position){
        if(position < 0 || position >= vec.Count) return;
        vec.RemoveAt(position);
    }

    public void RemoveList(List<char> vec, List<int> positions){
        foreach(int position in positions){
            if(position >= 0 && position < vec.Count)
                vec.RemoveAt(position);
        }
    }

    public void RemoveListCharRange(List<char> vec, int start, int off){
        if(start < 0 || start >= vec.Count) return;
        if(off < 0 || off >= vec.Count) return;
        if(start >= off) return;
        vec.RemoveRange(start, off);
    }
    
    // Convert Char to String
    // T1
    public List<string> GetListString(){
        List<string> vec1 = new List<string> { "aa","bb","cc"};
        vec1.Add("dd");
        return vec1;
    }

    public List<string> GetNCopyString(string ele, int size){
        List<string> vec = new List<string>(Enumerable.Repeat(ele, size));
        return vec;
    }

    // T2
    public void ReSizeListString(List<string> vec, int size, string ele){
        if (vec.Count >= size) {
            vec.RemoveRange(size, vec.Count);
        } else {
            vec.AddRange(Enumerable.Repeat(ele, size - vec.Count));
        }
    }

    // T3
    public void ClearSizeListString(List<string> vec, int size){
        if(vec.Count == 0) return;
        if(vec.Count <= size) {
            vec.Clear();
        }
    }

    // T4
    public int InsertListString(List<string> vec, int index, string ele){
        if(index > vec.Count || index < 0) return -1;
        vec.Insert(index, ele);
        return vec.Count;
    }

    // T5
    public int InsertListStringWithList(List<string> vec, int index, List<string> eles){
        if(index > vec.Count || index < 0) return -1;
        vec.InsertRange(index, eles);
        return vec.Count;
    }

    // T6
    public bool ContainsAnyString(List<string> vec, List<string> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(string ele in eles){
            if(vec.Contains(ele))
                return true;
        }
        return false;
    }

    public bool ContainsAllString(List<string> vec, List<string> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(string ele in eles){
            if(!vec.Contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void SetListString(List<string> vec, int position , string v){
        if(position < 0 || position >= vec.Count) return;
        else{
            vec[position] = v;
        }
    }

    public void SetListStringWithList(List<string> vec, List<int> position, List<string> v){
        if(vec.Count < position.Count) return;
        else{
            for(int i=0; i<position.Count; ++i){
                vec[position[i]] = v[i];
            }
        }
    }

    // T8
    public void RemoveListString(List<string> vec, int position){
        if(position < 0 || position >= vec.Count) return;
        vec.RemoveAt(position);
    }

    public void RemoveList(List<string> vec, List<int> positions){
        foreach(int position in positions){
            if(position >= 0 && position < vec.Count)
                vec.RemoveAt(position);
        }
    }

    public void RemoveListStringRange(List<string> vec, int start, int off){
        if(start < 0 || start >= vec.Count) return;
        if(off < 0 || off >= vec.Count) return;
        if(start >= off) return;
        vec.RemoveRange(start, off);
    }
    
    // Convert Char to Int
    // T1
    public List<int> GetListInt(){
        List<int> vec1 = new List<int> { 1, 2, 3};
        vec1.Add(4);
        return vec1;
    }

    public List<int> GetNCopyInt(int ele, int size){
        List<int> vec = new List<int>(Enumerable.Repeat(ele, size));
        return vec;
    }

    // T2
    public void ReSizeListInt(List<int> vec, int size, int ele){
        if (vec.Count >= size) {
            vec.RemoveRange(size, vec.Count);
        } else {
            vec.AddRange(Enumerable.Repeat(ele, size - vec.Count));
        }
    }

    // T3
    public void ClearSizeListInt(List<int> vec, int size){
        if(vec.Count == 0) return;
        if(vec.Count <= size) {
            vec.Clear();
        }
    }

    // T4
    public int InsertListInt(List<int> vec, int index, int ele){
        if(index > vec.Count || index < 0) return -1;
        vec.Insert(index, ele);
        return vec.Count;
    }

    // T5
    public int InsertListIntWithList(List<int> vec, int index, List<int> eles){
        if(index > vec.Count || index < 0) return -1;
        vec.InsertRange(index, eles);
        return vec.Count;
    }

    // T6
    public bool ContainsAnyInt(List<int> vec, List<int> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(int ele in eles){
            if(vec.Contains(ele))
                return true;
        }
        return false;
    }

    public bool ContainsAllInt(List<int> vec, List<int> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(int ele in eles){
            if(!vec.Contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void SetListInt(List<int> vec, int position , int v){
        if(position < 0 || position >= vec.Count) return;
        else{
            vec[position] = v;
        }
    }

    public void SetListIntWithList(List<int> vec, List<int> position, List<int> v){
        if(vec.Count < position.Count) return;
        else{
            for(int i=0; i<position.Count; ++i){
                vec[position[i]] = v[i];
            }
        }
    }

    // T8
    public void RemoveListInt(List<int> vec, int position){
        if(position < 0 || position >= vec.Count) return;
        vec.RemoveAt(position);
    }

    public void RemoveList(List<int> vec, List<int> positions){
        foreach(int position in positions){
            if(position >= 0 && position < vec.Count)
                vec.RemoveAt(position);
        }
    }

    public void RemoveListIntRange(List<int> vec, int start, int off){
        if(start < 0 || start >= vec.Count) return;
        if(off < 0 || off >= vec.Count) return;
        if(start >= off) return;
        vec.RemoveRange(start, off);
    }
    
    // Convert Char to Double
    // T1
    public List<double> GetListDouble(){
        List<double> vec1 = new List<double> { 1.0, 2.0, 3.0};
        vec1.Add(4.0);
        return vec1;
    }

    public List<double> GetNCopyDouble(double ele, int size){
        List<double> vec = new List<double>(Enumerable.Repeat(ele, size));
        return vec;
    }

    // T2
    public void ReSizeListDouble(List<double> vec, int size, double ele){
        if (vec.Count >= size) {
            vec.RemoveRange(size, vec.Count);
        } else {
            vec.AddRange(Enumerable.Repeat(ele, size - vec.Count));
        }
    }

    // T3
    public void ClearSizeListDouble(List<double> vec, int size){
        if(vec.Count == 0) return;
        if(vec.Count <= size) {
            vec.Clear();
        }
    }

    // T4
    public int InsertListDouble(List<double> vec, int index, double ele){
        if(index > vec.Count || index < 0) return -1;
        vec.Insert(index, ele);
        return vec.Count;
    }

    // T5
    public int InsertListDoubleWithList(List<double> vec, int index, List<double> eles){
        if(index > vec.Count || index < 0) return -1;
        vec.InsertRange(index, eles);
        return vec.Count;
    }

    // T6
    public bool ContainsAnyDouble(List<double> vec, List<double> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(double ele in eles){
            if(vec.Contains(ele))
                return true;
        }
        return false;
    }

    public bool ContainsAllDouble(List<double> vec, List<double> eles){
        if(vec.Count == 0 || eles.Count == 0) return false;
        foreach(double ele in eles){
            if(!vec.Contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void SetListDouble(List<double> vec, int position , double v){
        if(position < 0 || position >= vec.Count) return;
        else{
            vec[position] = v;
        }
    }

    public void SetListDoubleWithList(List<double> vec, List<int> position, List<double> v){
        if(vec.Count < position.Count) return;
        else{
            for(int i=0; i<position.Count; ++i){
                vec[position[i]] = v[i];
            }
        }
    }

    // T8
    public void RemoveListDouble(List<double> vec, int position){
        if(position < 0 || position >= vec.Count) return;
        vec.RemoveAt(position);
    }

    public void RemoveList(List<double> vec, List<int> positions){
        foreach(int position in positions){
            if(position >= 0 && position < vec.Count)
                vec.RemoveAt(position);
        }
    }

    public void RemoveListDoubleRange(List<double> vec, int start, int off){
        if(start < 0 || start >= vec.Count) return;
        if(off < 0 || off >= vec.Count) return;
        if(start >= off) return;
        vec.RemoveRange(start, off);
    }
}

File: Parallel-Java/src/parallel/containers/VectorParallel.java
package parallel.containers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class VectorParallel {
    // T1
    public ArrayList<Character> getListChar(){
        ArrayList<Character> vec1 = new ArrayList<>(Arrays.asList('a','b','c'));
        vec1.add('d');
        return vec1;
    }

    public ArrayList<Character> getNCopyChar(Character ele, int size){
        ArrayList<Character> vec = new ArrayList<>(Collections.nCopies(size, ele));
        return vec;
    }

    // T2
    public void reSizeListChar(ArrayList<Character> vec, int size, Character ele){
        if (vec.size() >= size) {
            vec.subList(size, vec.size()).clear();
        } else {
            vec.addAll(Collections.nCopies(size - vec.size(), ele));
        }
    }

    // T3
    public void clearSizeListChar(ArrayList<Character> vec, int size){
        if(vec.isEmpty()) return;
        if(vec.size() <= size) {
            vec.clear();
        }
    }

    // T4
    public int insertListChar(ArrayList<Character> vec, int index, Character ele){
        if(index > vec.size() || index < 0) return -1;
        vec.add(index, ele);
        return vec.size();
    }

    // T5
    public int insertListCharWithList(ArrayList<Character> vec, int index, List<Character> eles){
        if(index > vec.size() || index < 0) return -1;
        vec.addAll(index, eles);
        return vec.size();
    }

    // T6
    public boolean containsAnyChar(ArrayList<Character> vec, List<Character> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(Character ele : eles){
            if(vec.contains(ele))
                return true;
        }
        return false;
    }

    public boolean containsAllChar(ArrayList<Character> vec, List<Character> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(Character ele : eles){
            if(!vec.contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void setListChar(ArrayList<Character> vec, int position , Character v){
        if(position < 0 || position >= vec.size()) return;
        else{
            vec.set(position, v);
        }
    }

    public void setListCharWithList(ArrayList<Character> vec, ArrayList<Integer> position, ArrayList<Character> v){
        if(vec.size() < position.size()) return;
        else{
            for(int i=0; i<position.size(); ++i){
                vec.set(position.get(i), v.get(i));
            }
        }
    }

    // T8
    public void removeListChar(ArrayList<Character> vec, int position){
        if(position < 0 || position >= vec.size()) return;
        vec.remove(position);
    }

    public void removeListCharWithList(ArrayList<Character> vec, List<Integer> positions){
        for(Integer position : positions){
            if(position >= 0 && position < vec.size())
                vec.remove(position);
        }
    }

    public void removeListCharRange(ArrayList<Character> vec, int start, int off){
        if(start < 0 || start >= vec.size()) return;
        if(off < 0 || off >= vec.size()) return;
        if(start >= off) return;
        vec.subList(start, off).clear();
    }

    
    // Convert Char to String
    // T1
    public ArrayList<String> getListString(){
        ArrayList<String> vec1 = new ArrayList<>(Arrays.asList("aa","bb","cc"));
        vec1.add("dd");
        return vec1;
    }

    public ArrayList<String> getNCopyString(String ele, int size){
        ArrayList<String> vec = new ArrayList<>(Collections.nCopies(size, ele));
        return vec;
    }

    // T2
    public void reSizeListString(ArrayList<String> vec, int size, String ele){
        if (vec.size() >= size) {
            vec.subList(size, vec.size()).clear();
        } else {
            vec.addAll(Collections.nCopies(size - vec.size(), ele));
        }
    }

    // T3
    public void clearSizeListString(ArrayList<String> vec, int size){
        if(vec.isEmpty()) return;
        if(vec.size() <= size) {
            vec.clear();
        }
    }

    // T4
    public int insertListString(ArrayList<String> vec, int index, String ele){
        if(index > vec.size() || index < 0) return -1;
        vec.add(index, ele);
        return vec.size();
    }

    // T5
    public int insertListStringWithList(ArrayList<String> vec, int index, List<String> eles){
        if(index > vec.size() || index < 0) return -1;
        vec.addAll(index, eles);
        return vec.size();
    }

    // T6
    public boolean containsAnyString(ArrayList<String> vec, List<String> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(String ele : eles){
            if(vec.contains(ele))
                return true;
        }
        return false;
    }

    public boolean containsAllString(ArrayList<String> vec, List<String> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(String ele : eles){
            if(!vec.contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void setListString(ArrayList<String> vec, int position , String v){
        if(position < 0 || position >= vec.size()) return;
        else{
            vec.set(position, v);
        }
    }

    public void setListStringWithList(ArrayList<String> vec, ArrayList<Integer> position, ArrayList<String> v){
        if(vec.size() < position.size()) return;
        else{
            for(int i=0; i<position.size(); ++i){
                vec.set(position.get(i), v.get(i));
            }
        }
    }

    // T8
    public void removeListString(ArrayList<String> vec, int position){
        if(position < 0 || position >= vec.size()) return;
        vec.remove(position);
    }

    public void removeListStringWithList(ArrayList<String> vec, List<Integer> positions){
        for(Integer position : positions){
            if(position >= 0 && position < vec.size())
                vec.remove(position);
        }
    }

    public void removeListStringRange(ArrayList<String> vec, int start, int off){
        if(start < 0 || start >= vec.size()) return;
        if(off < 0 || off >= vec.size()) return;
        if(start >= off) return;
        vec.subList(start, off).clear();
    }


    // Convert Char to Integer
    // T1
    public ArrayList<Integer> getListInt(){
        ArrayList<Integer> vec1 = new ArrayList<>(Arrays.asList(1,2,3));
        vec1.add(4);
        return vec1;
    }

    public ArrayList<Integer> getNCopyInt(Integer ele, int size){
        ArrayList<Integer> vec = new ArrayList<>(Collections.nCopies(size, ele));
        return vec;
    }

    // T2
    public void reSizeListInt(ArrayList<Integer> vec, int size, Integer ele){
        if (vec.size() >= size) {
            vec.subList(size, vec.size()).clear();
        } else {
            vec.addAll(Collections.nCopies(size - vec.size(), ele));
        }
    }

    // T3
    public void clearSizeListInt(ArrayList<Integer> vec, int size){
        if(vec.isEmpty()) return;
        if(vec.size() <= size) {
            vec.clear();
        }
    }

    // T4
    public int insertListInt(ArrayList<Integer> vec, int index, Integer ele){
        if(index > vec.size() || index < 0) return -1;
        vec.add(index, ele);
        return vec.size();
    }

    // T5
    public int insertListIntWithList(ArrayList<Integer> vec, int index, List<Integer> eles){
        if(index > vec.size() || index < 0) return -1;
        vec.addAll(index, eles);
        return vec.size();
    }

    // T6
    public boolean containsAnyInt(ArrayList<Integer> vec, List<Integer> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(Integer ele : eles){
            if(vec.contains(ele))
                return true;
        }
        return false;
    }

    public boolean containsAllInt(ArrayList<Integer> vec, List<Integer> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(Integer ele : eles){
            if(!vec.contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void setListInt(ArrayList<Integer> vec, int position , Integer v){
        if(position < 0 || position >= vec.size()) return;
        else{
            vec.set(position, v);
        }
    }

    public void setListIntWithList(ArrayList<Integer> vec, ArrayList<Integer> position, ArrayList<Integer> v){
        if(vec.size() < position.size()) return;
        else{
            for(int i=0; i<position.size(); ++i){
                vec.set(position.get(i), v.get(i));
            }
        }
    }

    // T8
    public void removeListInt(ArrayList<Integer> vec, int position){
        if(position < 0 || position >= vec.size()) return;
        vec.remove(position);
    }

    public void removeListIntWithList(ArrayList<Integer> vec, List<Integer> positions){
        for(Integer position : positions){
            if(position >= 0 && position < vec.size())
                vec.remove(position);
        }
    }

    public void removeListIntRange(ArrayList<Integer> vec, int start, int off){
        if(start < 0 || start >= vec.size()) return;
        if(off < 0 || off >= vec.size()) return;
        if(start >= off) return;
        vec.subList(start, off).clear();
    }


    // Convert Char to Double
    // T1
    public ArrayList<Double> getListDouble(){
        ArrayList<Double> vec1 = new ArrayList<>(Arrays.asList(1.0,2.0,3.0));
        vec1.add(4.0);
        return vec1;
    }

    public ArrayList<Double> getNCopyDouble(Double ele, int size){
        ArrayList<Double> vec = new ArrayList<>(Collections.nCopies(size, ele));
        return vec;
    }

    // T2
    public void reSizeListDouble(ArrayList<Double> vec, int size, Double ele){
        if (vec.size() >= size) {
            vec.subList(size, vec.size()).clear();
        } else {
            vec.addAll(Collections.nCopies(size - vec.size(), ele));
        }
    }

    // T3
    public void clearSizeListDouble(ArrayList<Double> vec, int size){
        if(vec.isEmpty()) return;
        if(vec.size() <= size) {
            vec.clear();
        }
    }

    // T4
    public int insertListDouble(ArrayList<Double> vec, int index, Double ele){
        if(index > vec.size() || index < 0) return -1;
        vec.add(index, ele);
        return vec.size();
    }

    // T5
    public int insertListDoubleWithList(ArrayList<Double> vec, int index, List<Double> eles){
        if(index > vec.size() || index < 0) return -1;
        vec.addAll(index, eles);
        return vec.size();
    }

    // T6
    public boolean containsAnyDouble(ArrayList<Double> vec, List<Double> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(Double ele : eles){
            if(vec.contains(ele))
                return true;
        }
        return false;
    }

    public boolean containsAllDouble(ArrayList<Double> vec, List<Double> eles){
        if(vec.isEmpty() || eles.isEmpty()) return false;
        for(Double ele : eles){
            if(!vec.contains(ele))
                return false;
        }
        return true;
    }

    // T7
    public void setListDouble(ArrayList<Double> vec, int position , Double v){
        if(position < 0 || position >= vec.size()) return;
        else{
            vec.set(position, v);
        }
    }

    public void setListDoubleWithList(ArrayList<Double> vec, ArrayList<Integer> position, ArrayList<Double> v){
        if(vec.size() < position.size()) return;
        else{
            for(int i=0; i<position.size(); ++i){
                vec.set(position.get(i), v.get(i));
            }
        }
    }

    // T8
    public void removeListDouble(ArrayList<Double> vec, int position){
        if(position < 0 || position >= vec.size()) return;
        vec.remove(position);
    }

    public void removeListDoubleWithList(ArrayList<Double> vec, List<Integer> positions){
        for(Integer position : positions){
            if(position >= 0 && position < vec.size())
                vec.remove(position);
        }
    }

    public void removeListDoubleRange(ArrayList<Double> vec, int start, int off){
        if(start < 0 || start >= vec.size()) return;
        if(off < 0 || off >= vec.size()) return;
        if(start >= off) return;
        vec.subList(start, off).clear();
    }
}
